sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/ProactiveOpenConnectionsProcessor.java

```diff
@@ -0,0 +1,253 @@
+            if (taskList == null) {
+                taskList = new ArrayList<>();
+            }
+
+            taskList.add(openConnectionTask);
+            if (taskList.size() == 1) {
+                openConnectionsTaskSink.tryEmitNext(openConnectionTask);
+                openConnectionsTaskSinkBackUp.tryEmitNext(openConnectionTask);
+            }
+
+            return taskList;
+        });
+
+        RntbdEndpoint endpoint = this.endpointProvider.createIfAbsent(
+            openConnectionTask.getServiceEndpoint(),
+            openConnectionTask.getAddressUri().getURI(),
+            this,
+            openConnectionTask.getMinConnectionsRequiredForEndpoint());
+
+        endpoint.setMinChannelsRequired(Math.max(openConnectionTask.getMinConnectionsRequiredForEndpoint(), endpoint.getMinChannelsRequired()));
+    }
+
+    // There are two major kind tasks will be submitted
+    // 1. Tasks from up caller -> in this case, before we enqueue a task we would want to check whether the endpoint has already in the map
+    // 2. Tasks from existing flow -> for each endpoint, each time we will only 1 connection, if the connection count < the mini connection requirements,then re-queue.
+    // for this case, then there is no need to validate whether the endpoint exists
+    private synchronized void submitOpenConnectionWithinLoopInternal(OpenConnectionTask openConnectionTask) {
+        openConnectionsTaskSink.tryEmitNext(openConnectionTask);
+        openConnectionsTaskSinkBackUp.tryEmitNext(openConnectionTask);
+    }
+
+    @Override
+    public void close() throws IOException {
+        completeSink(openConnectionsTaskSink);
+        completeSink(openConnectionsTaskSinkBackUp);
+    }
+
+    public synchronized Disposable getOpenConnectionsPublisher() {
+
+        ConcurrencyConfiguration concurrencyConfiguration = concurrencySettings.get(aggressivenessHint.get());
+
+        return Flux.from(openConnectionsTaskSink.asFlux())
+            .publishOn(CosmosSchedulers.OPEN_CONNECTIONS_BOUNDED_ELASTIC)
+            .parallel(concurrencyConfiguration.openConnectionOperationEmissionConcurrency)
+            .runOn(CosmosSchedulers.OPEN_CONNECTIONS_BOUNDED_ELASTIC)
+            .flatMap(openConnectionTask -> {
+                Uri addressUri = openConnectionTask.getAddressUri();
+                URI serviceEndpoint = openConnectionTask.getServiceEndpoint();
+                int minConnectionsForEndpoint = openConnectionTask.getMinConnectionsRequiredForEndpoint();
+                String collectionRid = openConnectionTask.getCollectionRid();
+
+                return Flux.zip(Mono.just(openConnectionTask), openConnectionsHandler.openConnections(
+                        collectionRid,
+                        serviceEndpoint,
+                        Arrays.asList(addressUri),
+                        this,
+                        minConnectionsForEndpoint));
+            }, false, concurrencyConfiguration.openConnectionExecutionConcurrency)
```

Comment:
Since we are not delaying error here, I am curious what will happen if there is any error in this flux chain. Who handles it?

Reply:
Currently there is no handling - I rely on `OpenConnectionsHandler` to swallow errors. I think its a good idea to delay errors as much as possible especially in the non-blocking flow of opening connections we still have to continue opening in spite of errors pertaining to a particular inner sequence & have application traffic flow at the same time. Will pass `true` for `delayError` parameter.